### PR TITLE
Make inv(::Complex) return zero when input has Inf magnitude

### LINKED ISSUE #23134: Division of finite values by infinite complex values should always return zero
Since [Issue 9790](https://github.com/JuliaLang/julia/issues/9790) was closed and it was mentioned that corner cases should be separate issues, I am reposting one particular element of that issue where Julia is currently both wrong and, according to a comment in the thread, inconsistent with C++.

Division of any finite value by any complex infinite value should return 0. Currently, we have the following behavior in Julia 0.6.0:

    julia> 1/(Inf+im*Inf)
    NaN + NaN*im
    
    julia> 1/(Inf+im*NaN)
    NaN + NaN*im
    
    julia> 1/(Inf+0*im)
    0.0 - 0.0im

The third case is correct, but both of the other two cases are mathematically wrong. The first case is wrong. The second case is also wrong because if NaN is replaced by some real value, or by Inf, or by -Inf, it is clear that the result should be zero in each case. Further, it is inconsistent with the behavior of C++.

### PR DESCRIPTION
**Pull request by giordano**
Fix #23134.

### CODE REVIEW

**Comment by stevengj** [comment]
What is the performance impact of this?

**Comment by giordano** [comment]
Before the PR:
```julia
julia> v64 = randn(ComplexF64, 100000);

julia> @btime inv.($v64);
  1.393 ms (2 allocations: 1.53 MiB)

julia> v32 = randn(ComplexF32, 100000);

julia> @btime inv.($v32);
  426.785 μs (2 allocations: 781.33 KiB)

julia> vbig = big.(randn(ComplexF64, 1000));

julia> @btime inv.($vbig);
  1.057 ms (13001 allocations: 695.44 KiB)
```

After the PR:
```julia
julia> v64 = randn(ComplexF64, 100000);

julia> @btime inv.($v64);
  1.522 ms (2 allocations: 1.53 MiB)

julia> v32 = randn(ComplexF32, 100000);

julia> @btime inv.($v32);
  1.601 ms (2 allocations: 781.33 KiB)

julia> vbig = big.(randn(ComplexF64, 1000));

julia> @btime inv.($vbig);
  1.064 ms (13001 allocations: 695.44 KiB)
```

So, `ComplexF64` has a ~10% penalty, `Complex{BigFloat}` is basically unchanged, `ComplexF32` is much worse.  Note that for `ComplexF32` I changed the operation performed (before it was `conj(widen(z))/abs2(widen(z))`, in the PR it is `inv(widen(z))`).

Note also that as it is now, `ComplexF32` and `Complex{BigFloat}` give inconsistent results compared to `ComplexF64`:
```julia
julia> inv(complex(Inf32))
NaN32 - 0.0f0im

julia> inv(complex(big(Inf)))
NaN - 0.0im

julia> inv(complex(Inf))
0.0 - 0.0im
```

**Comment by stevengj** [comment]
With a big array, you are measuring cache and allocation time as well as computation.

File: base/complex.jl [hunk 1]
```diff
@@ -259,7 +259,11 @@ julia> conj(1 + 3im)
 conj(z::Complex) = Complex(real(z),-imag(z))
 abs(z::Complex)  = hypot(real(z), imag(z))
 abs2(z::Complex) = real(z)*real(z) + imag(z)*imag(z)
-inv(z::Complex)  = conj(z)/abs2(z)
+function inv(z::Complex)
+    a, b = reim(z)
+    (isinf(a) || isinf(b)) && return complex(copysign(zero(a), a), flipsign(-zero(b), b))
```
Comment (stevengj) [inline base/complex.jl]: `|` rather than `||` since we don't want short-circuit evaluation here (in the common case both will be `false`), I think?

Comment (stevengj) [reply base/complex.jl]: Also, I think the imaginary part should be:
```jl
flipsign(-copysign(zero(b), b), a)
```

Comment (giordano) [reply base/complex.jl]: In `(a + bi) / (c + di)`, the imaginary part is `(bc - ad) / (c^2 + d^2)`, so with `a = 1` and `b = 0` the numerator is simply `-d`, isn't it?  The numerator of the real part is `cd`, so probably that sign above is wrong?  `copysign(zero(a), sign(a) * sign(b)` instead?

Comment (stevengj) [reply base/complex.jl]: Whoops, right.

Comment (giordano) [reply base/complex.jl]: Ok, I changed `||` to `|` and left the rest as it was (apart from renaming `a` and `b` to `c` and `d` respectively), thanks!

**Comment by stevengj** [comment]
I tried
```
julia> a = rand(ComplexF64, 1000);

julia> f(x) = sum(inv, x)
f (generic function with 1 method)

julia> g(x) = sum(newinv, x)
g (generic function with 1 method)

julia> @btime f($a);
  14.100 μs (0 allocations: 0 bytes)

julia> @btime g($a);
  2.306 μs (0 allocations: 0 bytes)
```
so the new version is 7x faster?  Seems weird.

**Comment by giordano** [comment]
I don't know if that's the cause, but `conj(z)/abs2(z)` calls three times `real(z)` and other three times `imag(z)`, the new version calls `reim` only once

**Comment by stevengj** [comment]
It's maybe a change in whether it inlines?

**Comment by stevengj** [comment]
If I do
```jl
function newinv2(z::Complex)
    c, d = reim(z)
    complex(c, -d)/(c * c + d * d)
end
```
then `g2(x) = sum(newinv2, x)` gives
```jl
julia> @btime g2($a);
  1.125 μs (0 allocations: 0 bytes)
```
which is 2x faster than the version with the `isinf` checks.

**Comment by stevengj** [comment]
Oh, I just realized that I'm comparing the wrong version — I need the `inv(w::ComplexF64)`  version.

**Comment by chethega** [comment]
For Complex{Float32}, my machine counts a slowdown of 2.3 -> 8.5 cycles in batch / simd, and 6 -> 11 cycles for sequential processing. With `Complex{Float64}`, I count 7->8 and 7->10.5 cycles (on 2ghz broadwell). That's pretty expensive for a very rare corner case.
```
julia> using BenchmarkTools
julia> function _inv_new(z::Complex)
       c, d = reim(z)
       (isinf(c) | isinf(d)) && return complex(copysign(zero(c), c), flipsign(-zero(d), d))
       complex(c, -d)/(c * c + d * d)
       end
julia> _inv_old(z)=conj(z)/abs2(z);
julia> r=rand(Complex{Float32}, 1000); b=copy(r);
julia> @btime broadcast!($_inv_old, $b, $r); @btime broadcast!($_inv_new, $b, $r); @btime _inv_old($r[1]); @btime _inv_new($r[1]);
  1.143 μs (0 allocations: 0 bytes)
  4.268 μs (0 allocations: 0 bytes)
  3.075 ns (0 allocations: 0 bytes)
  5.506 ns (0 allocations: 0 bytes)
julia> r=rand(Complex{Float64}, 1000); b=copy(r);

julia> @btime broadcast!($_inv_old, $b, $r); @btime broadcast!($_inv_new, $b, $r); @btime _inv_old($r[1]); @btime _inv_new($r[1]);
  3.516 μs (0 allocations: 0 bytes)
  3.992 μs (0 allocations: 0 bytes)
  3.516 ns (0 allocations: 0 bytes)
  5.262 ns (0 allocations: 0 bytes)
```

**Comment by giordano** [comment]
I run your benchmark on my machine and I get very different results
```julia
julia> @btime broadcast!($_inv_old, $b, $r); @btime broadcast!($_inv_new, $b, $r); @btime _inv_old($r[1]); @btime _inv_new($r[1]);
  1.095 μs (0 allocations: 0 bytes)
  2.793 μs (0 allocations: 0 bytes)
  3.811 ns (0 allocations: 0 bytes)
  3.287 ns (0 allocations: 0 bytes)

julia> r=rand(Complex{Float64}, 1000); b=copy(r);

julia> @btime broadcast!($_inv_old, $b, $r); @btime broadcast!($_inv_new, $b, $r); @btime _inv_old($r[1]); @btime _inv_new($r[1]);
  3.810 μs (0 allocations: 0 bytes)
  3.811 μs (0 allocations: 0 bytes)
  3.812 ns (0 allocations: 0 bytes)
  3.815 ns (0 allocations: 0 bytes)
```
For `Float64` there is basically no change, for `Float32` the scalar benchmark seems to favour the new version

**Comment by stevengj** [comment]
@chethega, you made the same mistake I did: the `conj(z)/abs2(z)` is not the method that is called for `ComplexF32` or `ComplexF64`.   See the `inv(w::ComplexF64)`, which has a bunch of extra checks to handle overflow/underflow already.

**Comment by giordano** [comment]
Any other comment on this?

**Comment by StefanKarpinski** [comment]
@stevengj, please merge if this looks good to you.

**Pull request merged**